### PR TITLE
Add a new opt_lut pass, which combines inefficiently packed LUTs

### PR DESCRIPTION
**Pull request by whitequark**
On iCE40, techmapping `$alu` always results in a `SB_LUT4` cell with only 3 inputs used. This is inefficient and nothing later in the synthesis pipeline is currently able to take advantage of this unused LUT input. This LUT input could be very useful for certain ALU designs, as using it to fold a multiplexer LUT into an ALU LUT in an expression such as `c ? a : a + b` or `c ? 0 : a - b` could result, in extreme cases, a 2x reduction in LUT count.

The new `opt_lut` pass is able to combine *any* _k_-LUT whose output is connected solely to an _n_-LUT where both of them share _m_ inputs, given that at least one of these LUTs has a width of _k_ + _n_ - _m_ - 1. (Attributes can be used to restrict LUT combining, such as to preserve instantiated carry logic.)

This pass is very generic and operates on any `$lut` cells, even ones of different size. To work with iCE40's `SB_LUT4`, the `ice40_unlut` pass (#716) is used; normal techmapping takes care of mapping the resulting `$lut` cells back, as `opt_lut` takes care to not disturb existing inputs or unnecessarily resize LUT cells.

This pass is tested by using `equiv_induct` and a behavioral model of `SB_CARRY` to check equivalence of several ALU elements similar to real-world ones before and after `opt_lut`.

This concludes my work on #715.

### CODE REVIEW

File: passes/opt/opt_lut.cpp [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+	int lut_width = lut->getParam("\\WIDTH").as_int();
+	Const lut_table = lut->getParam("\\LUT");
+	int lut_index = 0;
+
+	for (int i = 0; i < lut_width; i++)
+	{
+		SigBit input = sigmap(lut_input[i]);
+		if (inputs.count(input))
+		{
+			lut_index |= inputs[input] << i;
```
Comment (swetland) [inline passes/opt/opt_lut.cpp]: Will this run into problems if lut_width is greater than the number of bits an in int?

Unclear if that's possible (are these ideal LUTs of any size or device LUTs which will be limited to small widths).

Comment (whitequark) [reply passes/opt/opt_lut.cpp]: Assuming 32-bit ints, that would be a LUT with 4 Gbit of configuration data.

I feel pretty safe.

**Comment by whitequark** [comment]
Doing more real-world testing. Statistics for another design, which generates VGA synchronization and also colorful squares:
<details>
<summary>Without -relut</summary>

```
   Number of cells:                378
     SB_CARRY                       34
     SB_DFF                          6
     SB_DFFE                        13
     SB_DFFESR                      49
     SB_DFFSR                       17
     SB_GB                           1
     SB_IO                          26
     SB_LUT4                       231
     SB_PLL40_CORE                   1
```
</details>
<details>
<summary>With -relut</summary>

```
   Number of cells:                346
     SB_CARRY                       34
     SB_DFF                          6
     SB_DFFE                        13
     SB_DFFESR                      49
     SB_DFFSR                       17
     SB_GB                           1
     SB_IO                          26
     SB_LUT4                       199
     SB_PLL40_CORE                   1
```
</details>
[Design](https://github.com/YosysHQ/yosys/files/2647377/xxx.zip)

**Comment by whitequark** [comment]
@daveshah1 There are some very obvious improvements missing in this pass. For example, I want to sort the LUT inputs (without some attribute on the LUT, of course) and of course rewrite truth tables such that opt_merge can find and combine identical LUTs that only differ in input order.

**Comment by cliffordwolf** [comment]
Note: I've renamed opt_lut.cpp to opt_lut.cc in 2d98db7.

**Comment by whitequark** [comment]
Ah, sorry, I haven't noticed.

**Comment by cliffordwolf** [comment]
No worries. Just letting you know so your don't run into unexpected merge conflicts on your side.

**Pull request merged**